packers/jws.go

```diff
@@ -0,0 +1,384 @@
+}
+
+// MediaTypeSignedMessage is media type for jws
+const MediaTypeSignedMessage iden3comm.MediaType = "application/iden3-signed-json"
+
+// DIDResolverHandlerFunc resolves did
+type DIDResolverHandlerFunc func(did string) (*verifiable.DIDDocument, error)
+
+// Resolve function is responsible to call provided handler for resolve did document
+func (f DIDResolverHandlerFunc) Resolve(did string) (*verifiable.DIDDocument, error) {
+	return f(did)
+}
+
+// SignerResolverHandlerFunc resolves signer
+type SignerResolverHandlerFunc func(kid string) (crypto.Signer, error)
+
+// Resolve function return signer by kid
+func (f SignerResolverHandlerFunc) Resolve(kid string) (crypto.Signer, error) {
+	return f(kid)
+}
+
+// JWSPacker is packer that use jws
+type JWSPacker struct {
+	didResolverHandler        DIDResolverHandlerFunc
+	signerResolverHandlerFunc SignerResolverHandlerFunc
+}
+
+// SigningParams packer parameters for jws generation
+type SigningParams struct {
+	SenderDID string
+	Alg       jwa.SignatureAlgorithm
+	KID       string
+	DIDDoc    *verifiable.DIDDocument
+	iden3comm.PackerParams
+}
+
+// NewSigningParams defines the signing parameters for jws generation
+func NewSigningParams() SigningParams {
+	return SigningParams{}
+}
+
+// Verify checks if signing params are valid
+func (s *SigningParams) Verify() error {
+	if s.Alg == "" {
+		return errors.New("alg is required for signing params")
+	}
+	if s.SenderDID == "" {
+		return errors.New("sender did is required for signing params")
+	}
+	return nil
+}
+
+// NewJWSPacker creates new jws packer instance
+func NewJWSPacker(
+	didResolverHandler DIDResolverHandlerFunc,
+	opaqueSignerResolverHandlerFunc SignerResolverHandlerFunc,
+	// opaqueVerifierResolverHandlerFunc OpaqueVerifierResolverHandlerFunc,
+) *JWSPacker {
+	return &JWSPacker{
+		didResolverHandler,
+		opaqueSignerResolverHandlerFunc,
+		// opaqueVerifierResolverHandlerFunc,
+	}
+}
+
+// Pack returns packed message to transport envelope with a zero knowledge proof in JWS full serialized format
+func (p *JWSPacker) Pack(
+	payload []byte, params iden3comm.PackerParams) ([]byte, error) {
+
+	signingParams, ok := params.(SigningParams)
+	if !ok {
+		return nil, errors.New("params must be SigningParams")
+	}
+	if err := signingParams.Verify(); err != nil {
+		return nil, err
+	}
+
+	bm := &iden3comm.BasicMessage{}
+	err := json.Unmarshal(payload, &bm)
+	if err != nil {
+		return nil, errors.Errorf("invalid message payload: %v", err)
+	}
+
+	if bm.From != signingParams.SenderDID {
+		return nil, errors.New("msg singer must me be msg sender")
+	}
+
+	didDoc := signingParams.DIDDoc
+	if didDoc == nil {
+		didDoc, err = p.didResolverHandler.Resolve(signingParams.SenderDID)
+		if err != nil {
+			return nil, errors.Errorf("resolve did failed: %v", err)
+		}
+	}
+
+	vm, err := lookupForKid(didDoc, signingParams.KID)
+	if err != nil {
+		return nil, err
+	}
+
+	var kid string
+	if k, ok := vm.PublicKeyJwk["kid"].(string); ok {
+		kid = k
```

Comment:
did you mention that we should use kid always as id of VM?

Reply:
In JS implementation yes. But the golang version can have search by both.